### PR TITLE
feat: migrated product edit page to spa

### PR DESCRIPTION
**Pull request by rudrodip**
Issue: #856

# Description

This PR migrates the Product Edit page from the legacy React server-side rendering approach to the Inertia.js SPA architecture.

### Changes

**Inertia Migration**
* Created new Inertia page (`pages/Products/Edit.tsx`) with full form state management using Inertia's `useForm` hook
* Updated `LinksController#edit` to render via Inertia with `edit_props`, `dropbox_app_key`, and `current_tab` props
* Deleted legacy ERB view (`app/views/links/edit.html.erb`) and pack file (`app/javascript/packs/product_edit.ts`)
* Removed `ProductEditPage` from SSR registration

**Server-Side Tab Routing**
* Added explicit routes for each tab (`/edit`, `/edit/content`, `/edit/receipt`, `/edit/share`) with `defaults: { tab: "..." }`
* Replaced React Router navigation with Inertia `Link` components using `preserveScroll` to prevent skeleton flashing on tab changes
* Added `edit_link_tab_path` helper to redirect back to the correct tab after form submission

**Form Handling**
* Implemented form submissions using `form.transform()` and `form.post()` with `preserveScroll: true` and `preserveState: true`
* Controller returns proper Inertia redirects with flash messages for success/warning/error states
* Passes `current_tab` in form data to maintain tab context through save operations

**Dropbox Integration**
* Introduced `useDropboxDropins` hook that lazily loads and caches the Dropbox script on-demand
* Exposes type-safe `choose`/`save` helpers and renders buttons only after script is ready
* Passed `dropbox_app_key` through Inertia props and exposed via `ProductEditContext`

**SPA Navigation**
* Updated `NewProductPage` to use `router.visit()` instead of `window.location.href` for seamless navigation to edit page after product creation

---

# Before

https://github.com/user-attachments/assets/a9675d18-1473-4496-bdcf-704becb3fe84

---

# After

https://github.com/user-attachments/assets/c36b7d8f-54ae-4214-b4d7-baa5c1cd3236

---

# Test Results

Have written e2e tests for this page and tested them

`bundle exec rspec spec/system/inertia_pages_spec.rb --format html --out rspec_results.html`

Screenshot 2026-01-07 at 11 45 06 PM

---

# Checklist

- [x] I have read the [contributing guidelines](https://github.com/antiwork/gumroad/blob/main/CONTRIBUTING.md)
- [x] I have watched [Gumroad PR review livestreams](https://www.youtube.com/@anti-work)
- [x] I have performed a self-review and left review comments on my PR
- [x] I have added/updated tests for my changes

---

# AI Disclosure

I used Claude Opus 4.5 via Cursor to find edge-cases, implement corner case solutions, and assist with the Inertia migration patterns. Self-reviewed all the code.

### CODE REVIEW

**Comment by rudrodip** [comment]
@Pradumn27 can you take a look please? thanks

**Comment by Pradumn27** [comment]
@Pradumn27 _a

File: app/controllers/links_controller.rb [hunk 1]
```diff
@@ -328,8 +328,15 @@ def edit
     redirect_to bundle_path(@product.external_id) if @product.is_bundle?
 
     @title = @product.name
-
     @presenter = ProductPresenter.new(product: @product, pundit_user:)
+    @current_tab = params[:tab] || "product"
```
Comment (Pradumn27) [inline app/controllers/links_controller.rb:332]: maintaining the current tab state is something to be computed on the frontend and not here, if it is used to highlight the selected tab then please see the patterns used in settings and wishlists in the codebase

File: app/controllers/links_controller.rb [hunk 2]
```diff
@@ -431,7 +438,12 @@ def update
       else
         error_message = @product.errors.full_messages.first || e.message
       end
-      return render json: { error_message: }, status: :unprocessable_entity
+      if request.inertia?
+        redirect_to edit_link_tab_path(@product.unique_permalink), inertia: { errors: { base: [error_message] } }, alert: error_message
```
Comment (Pradumn27) [inline app/controllers/links_controller.rb:442]: here we can simply always redirect to just the edit page on the first tab

File: app/controllers/links_controller.rb [hunk 3]
```diff
@@ -453,12 +465,21 @@ def update
...
       end
 
-      return render json: {
-        warning_message: "The following offer #{"code".pluralize(all_invalid_offer_codes.count)} #{issue_description}: #{all_invalid_offer_codes.join(", ")}. Please update #{all_invalid_offer_codes.length > 1 ? "them or they" : "it or it"} will not work at checkout."
-      }
+      warning_message = "The following offer #{"code".pluralize(all_invalid_offer_codes.count)} #{issue_description}: #{all_invalid_offer_codes.join(", ")}. Please update #{all_invalid_offer_codes.length > 1 ? "them or they" : "it or it"} will not work at checkout."
+
+      if request.inertia?
+        flash[:warning] = warning_message
+        return redirect_to edit_link_tab_path(@product.unique_permalink), status: :see_other
```
Comment (Pradumn27) [inline app/controllers/links_controller.rb:472]: directly pass the warning message as alert to redirect_to like you did above

File: app/javascript/components/ProductEdit/ContentTab/index.tsx [hunk 4]
```diff
@@ -92,8 +93,19 @@ export const extensions = (productId: string, extraExtensions: TiptapNode[] = []
...
+    id,
+    product,
+    updateProduct,
+    seller,
+    save,
+    existingFiles,
+    setExistingFiles,
+    uniquePermalink,
+    filesById,
+    dropboxAppKey,
```
Comment (Pradumn27) [inline app/javascript/components/ProductEdit/ContentTab/index.tsx:106]: please run linters and auto fix on the code

File: config/routes.rb [hunk 5]
```diff
@@ -685,8 +685,10 @@ def product_info_and_purchase_routes(named_routes: true)
 
     # TODO: move these within resources :products block above
     get "/products/paged", to: "links#products_paged", as: :products_paged
-    get "/products/:id/edit", to: "links#edit", as: :edit_link
-    get "/products/:id/edit/*other", to: "links#edit"
+    get "/products/:id/edit", to: "links#edit", as: :edit_link, defaults: { tab: "product" }
+    get "/products/:id/edit/content", to: "links#edit", as: :edit_link_content, defaults: { tab: "content" }
+    get "/products/:id/edit/receipt", to: "links#edit", as: :edit_link_receipt, defaults: { tab: "receipt" }
+    get "/products/:id/edit/share", to: "links#edit", as: :edit_link_share, defaults: { tab: "share" }
```
Comment (Pradumn27) [inline config/routes.rb:691]: why do we need these defaults here?

File: app/javascript/pages/Products/Edit.tsx
Comment (Pradumn27) [inline app/javascript/pages/Products/Edit.tsx:1]: every product tab needs to be a separate page which is rendered by the backend thus they should not be coupled here in one component rather structured to be separate and coupled via the layout and provider states

File: app/javascript/pages/Products/Edit.tsx [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+        onError: (errors) => {
+          const errorMessage = Object.values(errors)[0] || "Something went wrong";
+          showAlert(errorMessage, "error");
+          reject(new Error(errorMessage));
+        },
+      });
+    });
+  };
+
+  const contextValue = React.useMemo(
```
Comment (Pradumn27) [inline app/javascript/pages/Products/Edit.tsx:188]: create a separate state provider to be used by all the tabs

File: app/javascript/pages/Products/Edit.tsx [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+
+  const updateProduct = (update: Partial<Product> | ((product: Product) => void)) =>
+    setProduct((prevProduct) => {
+      const updated = { ...prevProduct };
+      if (typeof update === "function") update(updated);
+      else Object.assign(updated, update);
+      return updated;
+    });
+
+  // Inertia form for submissions
```
Comment (Pradumn27) [inline app/javascript/pages/Products/Edit.tsx:106]: Don't add comments unless necessary to justify what's happening and code is not self explanatory

File: app/javascript/pages/Products/Edit.tsx [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+        newlyAdded ? { ...variant, id: null } : variant,
+      ),
+      availabilities: filteredProduct.availabilities.map(({ newlyAdded, ...availability }) =>
+        newlyAdded ? { ...availability, id: null } : availability,
+      ),
+      installment_plan: filteredProduct.allow_installment_plan ? filteredProduct.installment_plan : null,
+      current_tab,
+    }));
+
+    return new Promise<void>((resolve, reject) => {
```
Comment (Pradumn27) [inline app/javascript/pages/Products/Edit.tsx:148]: why do we need this to be wrapped in a promise?

File: spec/system/inertia_pages_spec.rb
Comment (Pradumn27) [inline spec/system/inertia_pages_spec.rb:1]: no, we dont' want the specs to be coupled here, we have separate spec files for controllers, services etc, see which tests are to be changed/updated there and do that, add any tests in the particular spec file where the behaviour has updated or needs to be tested

File: rspec_results.html
Comment (Pradumn27) [inline rspec_results.html:1]: don't commit unnecessary files

File: app/javascript/components/ProductEdit/Layout.tsx [hunk 9]
```diff
@@ -131,36 +132,32 @@ export const Layout = ({
...
-  const tab = match?.handle ?? "product";
-
-  const navigate = useRefToLatest(useNavigate());
-
   const [isPublishing, setIsPublishing] = React.useState(false);
   const setPublished = async (published: boolean) => {
     try {
       setIsPublishing(true);
-      await saveProduct(uniquePermalink, id, product, currencyType);
+      await save();
```
Comment (Pradumn27) [inline app/javascript/components/ProductEdit/Layout.tsx:151]: what we want is to remove the API call being made here and it to be done via inertia instead, so we can partially load only the needed props

Comment (Pradumn27) [reply app/javascript/components/ProductEdit/Layout.tsx:151]: also delete any files/function calls that you remove if they are not used anywhere else

**Comment by Pradumn27** [comment]
@EmCousin please close the PR, @rudrodip can open another one once the architectural issues and comments are resolved, thanks!
